core/modules/replica/ServiceManagementRequest.h

```diff
@@ -164,8 +164,19 @@ class ServiceManagementRequest
      * @see Request::notifyImpl()
      */
     void notifyImpl() final {
-        if (_onFinish) {
-            _onFinish(shared_from_base<ServiceManagementRequest<POLICY>>());
+
+        if (nullptr != _onFinish) {
+    
+            // Clearing the stored callback after finishing the up-stream notification
+            // has two purposes:
+            //
+            // 1. it guaranties (exactly) one time notification
```

Comment:
guarantees.

I've seen this comment few times already, maybe you should refactor this into a separate class/method?